File: python-algorithm/leetcode/problem_968.py
"""968. Binary Tree Cameras
https://leetcode.com/problems/binary-tree-cameras/

Given a binary tree, we install cameras on the nodes of the tree. 

Each camera at a node can monitor its parent, itself, and its immediate children.

Calculate the minimum number of cameras needed to monitor all nodes of the tree.

Example 1:
    <>
   /
  <O>
  /\
<>  <>
Input: [0,0,null,0,0]
Output: 1
Explanation: One camera is enough to monitor all nodes if placed as shown.

Input: [0,0,null,0,null,0,null,null,0]
       <>
      /
     <O>
     /
    <>
    /
   <O>
    \
    <>
Output: 2
Explanation: At least two cameras are needed to monitor all nodes of the tree.
The above image shows one of the valid configurations of camera placement.

Note:

The number of nodes in the given tree will be in the range [1, 1000].
Every node has value 0.
"""
from common.tree_node import TreeNode


class Solution:
    def min_camera_cover(self, root: TreeNode) -> int:
        return 0

File: java-algorithm/src/main/java/com/leetcode/Problem968.java
package com.leetcode;

import com.common.TreeNode;

/**
 * 968. Binary Tree Cameras
 * https://leetcode.com/problems/binary-tree-cameras/
 */
public class Problem968 {
    private int cnt;

    public int minCameraCover(TreeNode root) {
        if (dfs(root) < 1) {
            cnt++;
        }
        return cnt;
    }

    private int dfs(TreeNode node) {
        if (node == null) {
            return 2;  //
        }
        int left = dfs(node.left);
        int right = dfs(node.right);
        if (left == 0 || right == 0) {
            cnt++;
            return 1;
        }
        return left == 1 || right == 1 ? 2 : 0;
    }
}
